### ISSUE TITLE
[No issue] Explanation request regarding BMC trace print

### ISSUE THREAD

**Issue by Robbin1987** [seahorn/seahorn#80]
Hi All,

Could not find a better area to post this request for information. What I would like to know what is how to understand the values given by the BMC trace.

For example (relates to the code below):
 ` %12 = call i32 @ValidateFunc() #3
  %13 = icmp eq i32 %12, 0'

%12 has the value of 1, which means that %13 get the value FALSE.
Looking at the BMC trace, %13, has a value of FALSE written behind it. This seems clear.

What I don't get is the value behind: %10 or %11. I would expect the result of add(1,1) = 2, but that is not the case.

Full code snippet with all convertions:
`			int x=add(1,1); int y=add(2,2);
 if(ValidateFunc()) {
   int t1 = x;
   int t2 = y;
   x = t1+ t2;
   y = t1 + t2;
  }
   sassert(y < 1);`

The code snippet has the following IR code after some pp:
`precall6:                                         ; preds = %precheck
  br label %count

count:                                            ; preds = %precall6
  call void @seahorn.fn.enter() #3
  %10 = call fastcc i32 @add(i32 1, i32 1) #3
  %11 = call fastcc i32 @add(i32 2, i32 2) #3
  %12 = call i32 @ValidateFunc() #3
  %13 = icmp eq i32 %12, 0
  %14 = select i1 %13, i32 0, i32 %10
  %..i = add nsw i32 %11, %14
  %15 = icmp slt i32 %..i, 1
  call void @verifier.assume.not(i1 %15)
  %16 = xor i1 %15, true
  call void @llvm.assume(i1 %16), !seahorn !1
  br label %17

; <label>:17                                      ; preds = %count
  br label %verifier.error

verifier.error:                                   ; preds = %17
  call void @seahorn.fail()
  ret i32 42`

The BMC trace has the following result when generating the CEX:
`precall6: 
count: 
  %_10 0
  %_11 1
  %_12 1
  %_13 false
  %_14 0
  %..i 1
  %_15 false
  %_16 true
_bb: 
verifier.error: 
verifier.error.split: `

Thanks in advance,

**Comment by agurfinkel** [comment]
I am wondering wether you are looking at BMC trace without inlining. Unfortunately, we currently only support trace generation with inlining. When functions are not inlined, they are treated non-deterministically during trace generation leading to somewhat strange results.

**Comment by Robbin1987** [comment]
During the arguments experimentation I left out the seaopt completely, while the other commands have arguments that are similar to the default ones.

I will perform the seaopt again and see if I can use the BMC trace.

Thanks

**Comment by caballa** [comment]
You will need to add option `--inline`. This option will inline all functions assuming they are not recursive.

**Comment by Robbin1987** [comment]
I performed the default command with arguments to get the CEX.

CMD: sea pf sample_f.c --cex=harness.ll

BMC trace output:
`sat
entry: 
  %_0 1
  %_1 false
  %_2 true
  %_3 false
  %_4 true
  %_5 false
  %_6 true
  %_7 false
  %_8 true
  %_9 false
  %_10 true
  %_11 0
  %_12 0
  %_13 0
  %_14 1
  %_15 1
  %_16 false
  %_17 0
  %..i 1
  %_18 false
  %_19 true
entry.split: 
`
I expected that the value behind %_xx would be the result of IR instructions, but that is not the case:
`%11 = tail call i32 bitcast (i32 (...)* @add to i32 (i32, i32)*)(i32 1, i32 1) #3
  %12 = tail call i32 bitcast (i32 (...)* @add to i32 (i32, i32)*)(i32 2, i32 2) #3
  `
If possible I prefer not to inline, because that messes up the trace to the area of interest. 

Thanks for helping out!

**Comment by Robbin1987** [comment]
Hi Guys,

I noticed the closed ticket regarding the -g -inline issue. If I added these arguments:
 ./sea pf -g -inline /media/sf_seahorn/sample/sample_f.c --cex==/media/sf_seahorn/sample/test.ll

Then I get the result that I expect:
`sat
entry: 
enter: main
enter: addafter
  num = i32 2
  loc = i32 2
  test = i32 4
  test = i32 4
enter: check
  x = i32 1
  y = i32 2
  t1 = i32 1
  t2 = i32 2
  z = i32 0
  x = i32 3
  y = i32 3
  z = i32 6
  test2 = i32 6
`
The values are also what I expect. The BMC trace is even better than I hoped for :)!

For some reason, I suddenly get an LLVM ERROR: IO failure on output stream. This was independent of the -g or -inline arguments. It results in not producing a CEX.
